Introduce `AnimationStyle`

TahaTesser opened the pull request:
This PR introduces `AnimationStyle`, it is used to override default animation curves and durations in several widgets.

fixes  [Add the ability to customize MaterialApp theme animation duration](https://github.com/flutter/flutter/issues/78372)
fixes [Allow customization of showMenu transition animation curves and duration](https://github.com/flutter/flutter/issues/135638)


Here is an example where popup menu curve and transition duration is overriden:

```dart
          popUpAnimationStyle: AnimationStyle(
            curve: Easing.emphasizedAccelerate,
            duration: Durations.medium4,
          ),
```

Set `AnimationStyle.noAnimation` to disable animation.
```dart
    return MaterialApp(
      themeAnimationStyle: AnimationStyle.noAnimation,
```



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

TahaTesser committed:
Introduce `AnimationStyle`

TahaTesser commented:
@HansMuller 
Please take a lolok at this Animation Override draft

In the example, it would be like this

```dart
      body: Center(
        child: PopupMenuButton<SampleItem>(
          openAnimationOverride: () {
            return (Easing.emphasizedAccelerate, const Duration(milliseconds: 2));
          },
          closeAnimationOverride: () {
            return (Easing.legacyDecelerate, const Duration(milliseconds: 2));
          },
```


### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

/// Flutter code sample for [PopupMenuButton].

// This is the type used by the popup menu below.
enum SampleItem { itemOne, itemTwo, itemThree }

void main(){
  // timeDilation = 5;
  runApp(const PopupMenuApp());
}

class PopupMenuApp extends StatelessWidget {
  const PopupMenuApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: PopupMenuExample(),
    );
  }
}

class PopupMenuExample extends StatefulWidget {
  const PopupMenuExample({super.key});

  @override
  State<PopupMenuExample> createState() => _PopupMenuExampleState();
}

class _PopupMenuExampleState extends State<PopupMenuExample>  {
  SampleItem? selectedMenu;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('PopupMenuButton')),
      body: Center(
        child: PopupMenuButton<SampleItem>(
          openAnimationOverride: () {
            return (Easing.emphasizedAccelerate, const Duration(milliseconds: 2));
          },
          closeAnimationOverride: () {
            return (Easing.legacyDecelerate, const Duration(milliseconds: 2));
          },
          initialValue: selectedMenu,
          // Callback that sets the selected popup menu item.
          onSelected: (SampleItem item) {
            setState(() {
              selectedMenu = item;
            });
          },
          itemBuilder: (BuildContext context) => <PopupMenuEntry<SampleItem>>[
            const PopupMenuItem<SampleItem>(
              value: SampleItem.itemOne,
              child: Text('Item 1'),
            ),
            const PopupMenuItem<SampleItem>(
              value: SampleItem.itemTwo,
              child: Text('Item 2'),
            ),
            const PopupMenuItem<SampleItem>(
              value: SampleItem.itemThree,
              child: Text('Item 3'),
            ),
          ],
        ),
      ),
    );
  }
}

```

</details>

HansMuller commented:
What if I only wanted to override the curve or duration? The implication here is that I'd always have to override all of them. What if we created a class that was more like a ThemeData:

```dart
class AnimationOverride {
  const AnimationOverride({
    this.curve,
    this.reverseCurve,
    this.duration,
  });

  final Curve? curve;
  final Curve? reverseCurve;
  final Duration? duration;
}
```

TahaTesser commented:
How do you suggest we pass these values from the widget? This still adds multiple parameters

TahaTesser commented:
@hans
Updated the animation override class and provide it this `PopupMenuButton` and `MaterialApp` to start.

I think just `animationOverride` is enough for the `PopupMenuButton`, instead of open and close. It seems to be reverse the same animation. 


I can see some see have `duration` and some with route have `transitionDuration` and `reverseTransitionDuration (like modal sheet). Do we add `transitionDuration`and `reverseTransitionDuration` parameters to the class or just `duration` or `reverseDuration` is enough to override such properties?

HansMuller commented:
I think just having duration and reverseDuration is enough. Adding a transitionDuration parameter would make the AnimationOverrides class a little confusing.

HansMuller commented on examples/api/lib/material/app/app.0.dart:
```diff
@@ -0,0 +1,71 @@
+      // When the button with `Customize Theme Animation` text is pressed,
+      // the theme animation will be overridden with the following
+      // [AnimationOverride] instance in the [MaterialApp].
+      themeAnimationOverride: customizedThemeAnimation ? AnimationOverride(
+        curve: Easing.emphasizedDecelerate,
+        duration: const Duration(seconds: 3),
+      ) : null,
```
Two trivial suggestions: rename customizedThemeAnimation to isCustomizedThemeAnimation so that it sounds more like a flag than an instance, and put the null `?:` case first to make the expression a tiny bit easier to read.

Same suggestion for the popup menu example.

```suggestion
themeAnimationOverride: !isCustomizedThemeAnimation ? null : AnimationOverride(
  curve: Easing.emphasizedDecelerate,
  duration: const Duration(seconds: 3),
),
```

HansMuller commented on examples/api/lib/material/popup_menu/popup_menu.2.dart:
```diff
@@ -0,0 +1,131 @@
+}
+
+class _PopupMenuExampleState extends State<PopupMenuExample> {
+  SampleItem? selectedMenu;
```
Looks like this should be called selectedItem

HansMuller commented on examples/api/lib/material/popup_menu/popup_menu.2.dart:
```diff
@@ -0,0 +1,131 @@
+class _PopupMenuExampleState extends State<PopupMenuExample> {
+  SampleItem? selectedMenu;
+  bool customizedThemeAnimation = false;
+  bool slowAnimation = false;
```
Same comment as earlier; `isSlowAnimation` sounds more like a flag.

HansMuller commented on examples/api/lib/material/popup_menu/popup_menu.2.dart:
```diff
@@ -0,0 +1,131 @@
+                  ) : null,
+                  icon: const Icon(Icons.more_vert),
+                  initialValue: selectedMenu,
+                  // Callback that sets the selected popup menu item.
```
This comment isn't necessary. Even less so if the body of the setState is `selectedMenuItem = item` :-)

HansMuller commented on packages/flutter/lib/src/animation/animation_override.dart:
```diff
@@ -0,0 +1,36 @@
+
+import 'curves.dart';
+
+/// A class that can be used to override the default animations of a widget.
```
```suggestion
/// Used to override the default parameters of an animation.
```

HansMuller commented on packages/flutter/lib/src/animation/animation_override.dart:
```diff
@@ -0,0 +1,36 @@
+/// - [MaterialApp]
+/// - [PopupMenuButton]
+///
+/// All of the parameters are optional. If no parameters are specified,
```
I think it would be useful to claim that if the duration and reverseDuration properties the corresponding animation will have been disabled.

TahaTesser replied:
Do you mean adding a line that says this?

```dart
/// If [duration] and [reverseDuration] are set to [Duration.zero], the
/// corresponding animation will be disabled.
```

HansMuller replied:
Yes!

HansMuller commented on examples/api/lib/material/app/app.0.dart:
```diff
@@ -0,0 +1,71 @@
+      // When the button with `Customize Theme Animation` text is pressed,
+      // the theme animation will be overridden with the following
+      // [AnimationOverride] instance in the [MaterialApp].
+      themeAnimationOverride: customizedThemeAnimation ? AnimationOverride(
```
A request that comes up now and then is for a way to prevent an animation from happening at all. Perhaps we should provide a constant like AnimationOverride.noAnimation that has the durations set to zero. This feature would be useful to include here, since that's sometimes what developers want to do to the theme animation.

TahaTesser replied:
Great idea!

HansMuller commented on packages/flutter/lib/src/material/app.dart:
```diff
@@ -753,6 +763,22 @@ class MaterialApp extends StatefulWidget {
+  /// the theme animation curve in the underlying [AnimatedTheme] widget.
+  /// If it is null, then [themeAnimationCurve] will be used. Otherwise,
+  /// defaults to [Curves.linear].
+  final AnimationOverride? themeAnimationOverride;
```
Maybe we should also include a pointer to the example here.

HansMuller commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -1292,6 +1305,24 @@ class PopupMenuButton<T> extends StatefulWidget {
   /// Defaults to [Clip.none].
   final Clip clipBehavior;
 
+  /// When an instance of [AnimationOverride] is provided, this will be used to
```
```suggestion
Used to override the default animation curves and durations for the popup menu's open and close transitions.
```

HansMuller commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -1292,6 +1305,24 @@ class PopupMenuButton<T> extends StatefulWidget {
+  ///
+  /// If [AnimationOverride.reverseCurve] is provided, it will be used to
+  /// override the default popup animation reverse curve. Otherwise, defaults to
+  /// `const Interval(0.0, _kMenuCloseIntervalEnd)` where `_kMenuCloseIntervalEnd`
```
```suggestion
  /// `Interval(0.0, 2.0 / 3.0)`
```

HansMuller commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -1292,6 +1305,24 @@ class PopupMenuButton<T> extends StatefulWidget {
+  /// has a value of `2.0 / 3.0`.
+  ///
+  /// If [AnimationOverride.duration] is provided, it will be used to override
+  /// the default popup animation duration. Otherwise, defaults to `_kMenuDuration`
```
I don't think we should document private variables. It's OK to just say 300ms.

TahaTesser commented:
@HansMuller appreciate the review!

TahaTesser commented:
Added the suggestions and updated PR description

HansMuller commented:
We already have TextStyle, MenuStyle, ButtonStyle, all classes that are just bundles of properties that modify the appearance of a class. Maybe we should call this class `AnimationStyle` instead of `AnimationOverrides`.

TahaTesser commented on packages/flutter/test/material/app_test.dart:
```diff
@@ -1525,6 +1525,71 @@ void main() {
+    expect(tester.widget<Material>(find.byType(Material)).color, const Color(0xff1c1b1f));
+
+    // Switch from dark to light theme with overriden animation duration.
+    await tester.pumpWidget(buildWidget(animationStyle: AnimationStyle.noAnimation));
```
Replace `Duration.zero` with `AnimationStyle.noAnimation` :)

TahaTesser commented:
Resolved conflict from another PR just landed today and rename class to `AnimationStyle`

HansMuller commented on packages/flutter/lib/src/material/app.dart:
```diff
@@ -937,8 +962,8 @@ class _MaterialAppState extends State<MaterialApp> {
         cursorColor: effectiveCursorColor,
         child: AnimatedTheme(
```
If `widget.themeAnimationStyle` is `noAnimation` could we just leave the AnimatedTheme out?

HansMuller commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -834,18 +835,19 @@ class _PopupMenuRoute<T> extends PopupRoute<T> {
+  final AnimationStyle? popUpAnimationStyle;
 
   @override
   Animation<double> createAnimation() {
```
Similar to MaterialApp: if popupAnimationStyle is noAnimation, can we leave the CurvedAnimation out?

HansMuller commented on examples/api/lib/material/app/app.0.dart:
```diff
@@ -0,0 +1,68 @@
+        seedColor: Colors.green,
+        brightness: Brightness.dark,
+      )),
+      themeAnimationStyle: !isCustomizedThemeAnimation ? null : AnimationStyle(
```
A segmented button labeled AnimationStyle: with three options: `default, custom, none` would also show devs how to turn the animation off (a feature that's been asked for)

HansMuller commented on examples/api/lib/material/popup_menu/popup_menu.2.dart:
```diff
@@ -0,0 +1,130 @@
+              mainAxisSize: MainAxisSize.min,
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: <Widget>[
+                OutlinedButton(
```
Same comment about as for the MaterialApp demo: maybe show three options here.

HansMuller commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -1299,6 +1311,22 @@ class PopupMenuButton<T> extends StatefulWidget {
+  /// If [AnimationStyle.reverseCurve] is provided, it will be used to
+  /// override the default popup animation reverse curve. Otherwise, defaults to
+  /// `Interval(0.0, 2.0 / 3.0)`.
+  ///
```
Also mention:   `/// To disable the theme animation, use [AnimationStyle.noAnimation].`

HansMuller commented on packages/flutter/lib/src/material/app.dart:
```diff
@@ -753,6 +756,28 @@ class MaterialApp extends StatefulWidget {
+  /// If it is null, then [themeAnimationCurve] will be used. Otherwise,
+  /// defaults to [Curves.linear].
+  ///
+  /// To disable the theme animation, set [AnimationStyle.noAnimation].
```
```suggestion
  /// To disable the theme animation, use [AnimationStyle.noAnimation].
```

HansMuller reviewed (commented):
This is looking good. Only significant comment is about actually eliminating the animation widgets for `AnimationStyle.noAnimation`

TahaTesser commented:
@HansMuller 

> This is looking good. Only significant comment is about actually eliminating the animation widgets for `AnimationStyle.noAnimation`

Updated the code for `AnimationStyle.noAnimation`, updated examples and tests for `SegmentedButton`, and documentation

HansMuller commented on packages/flutter/lib/src/material/app.dart:
```diff
@@ -930,33 +955,44 @@ class _MaterialAppState extends State<MaterialApp> {
+    Widget childWidget = widget.builder != null
+      ? Builder(
+          builder: (BuildContext context) {
+            // Why are we surrounding a builder with a builder?
```
This comment needs to be updated since the AnimatedTheme is no longer "above" and because the AnimatedTheme is optional.

Do we even need the extra build if the AnimatedTheme isn't used?

TahaTesser replied:
Woah... good catch I thought i did remove `AnimatedTheme`.

HansMuller reviewed (commented):
This appears to be just about complete. Just a little feedback about Theme and AnimatedTheme

HansMuller reviewed (approved):
LGTM

HansMuller commented:
@TahaTesser - I had to revert this due to an internal name collision. 

When we reland this, we should include copyWith, hashCode and the changes from https://github.com/flutter/flutter/pull/138619

TahaTesser commented:
> @TahaTesser - I had to revert this (https://github.com/flutter/flutter/pull/138628) due to an internal name collision. 
> 
> 
> 
> When we reland this, we should include copyWith, hashCode methods, and the changes from https://github.com/flutter/flutter/pull/138619

Sounds good. Do we also change the name?
